src/github-handler/comment-handler/suggestion-patch-handler/scope-suggestion-hunks-handler.ts

```diff
@@ -0,0 +1,163 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {Hunk, Range} from '../../../types';
+
+/**
+ * Given a value and a list of ranges, find the index of the range domain
+ * that the value lies within.
+ * Return -1 otherwise.
+ * We expect the ranges to be ascending, and disjoint.
+ * @param {Range[]} ranges a list of range objects with a start value and end value
+ * @param {number} value the value to search for in the range list
+ * @param {number} from the lower bound of the search indices. (optional)
+ */
+export function findRange(ranges: Range[], value: number, from = 0): number {
+  function binarySearch(lo: number, hi: number): number {
+    const mid = Math.floor((hi + lo) / 2);
+    if (lo > hi) {
+      return -1;
+    } else if (ranges[mid].start > value) {
+      return binarySearch(lo, mid - 1);
+    } else if (ranges[mid].end < value) {
+      return binarySearch(mid + 1, hi);
+    } else {
+      return mid;
+    }
+  }
+  return binarySearch(from, ranges.length - 1);
+}
+
+/**
+ * Get a subset of the hunks who exactly fit within a valid hunk scope.
+ * Also get the converse of the set of hunks that are invalid.
+ * @param {Range[]} scope the list of valid ranges that a hunk can fit between
+ * @param {Hunk[]} suggestedHunks the hunks for the file to suggest
+ */
+function getValidSuggestionHunks(
+  scope: Range[],
+  suggestedHunks: Hunk[]
+): {inScopeHunks: Hunk[]; outOfScopeHunks: Hunk[]} {
```

Comment:
If both of these are already sorted, we should be able to do this in O(n + m) time rather than O(n log m) or O(m log n).

Depending on the expected sizes of the inputs, we might want to switch.

This is non-blocking for this PR.

Reply:
This would be really interesting to tackle as there'd probably need to be a lot of testing to see what the average case is for users.
https://github.com/googleapis/code-suggester/issues/87